sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java

```diff
@@ -0,0 +1,286 @@
+                return handleBodyReturnType(response, methodParser, bodyType)
+                    .map(bodyAsObject -> createResponse(response, entityType, bodyAsObject))
+                    .switchIfEmpty(Mono.fromCallable(() -> createResponse(response, entityType, null)));
+            }
+        } else {
+            // For now, we're just throwing if the Maybe didn't emit a value.
+            return handleBodyReturnType(response, methodParser, entityType);
+        }
+    }
+
+    private Mono<?> handleBodyReturnType(final HttpResponseDecoder.HttpDecodedResponse response,
+                                         final SwaggerMethodParser methodParser, final Type entityType) {
+        final int responseStatusCode = response.getSourceResponse().getStatusCode();
+        final HttpMethod httpMethod = methodParser.getHttpMethod();
+        final Type returnValueWireType = methodParser.getReturnValueWireType();
+
+        final Mono<?> asyncResult;
+        if (httpMethod == HttpMethod.HEAD
+            && (TypeUtil.isTypeOrSubTypeOf(entityType, Boolean.TYPE)
+            || TypeUtil.isTypeOrSubTypeOf(entityType, Boolean.class))) {
+            boolean isSuccess = (responseStatusCode / 100) == 2;
+            asyncResult = Mono.just(isSuccess);
+        } else if (TypeUtil.isTypeOrSubTypeOf(entityType, byte[].class)) {
+            // Mono<byte[]>
+            Mono<byte[]> responseBodyBytesAsync = response.getSourceResponse().getBodyAsByteArray();
+            if (returnValueWireType == Base64Url.class) {
+                // Mono<Base64Url>
+                responseBodyBytesAsync = responseBodyBytesAsync
+                    .mapNotNull(base64UrlBytes -> new Base64Url(base64UrlBytes).decodedBytes());
+            }
+            asyncResult = responseBodyBytesAsync;
+        } else if (FluxUtil.isFluxByteBuffer(entityType)) {
+            // Mono<Flux<ByteBuffer>>
+            asyncResult = Mono.just(response.getSourceResponse().getBody());
+        } else if (TypeUtil.isTypeOrSubTypeOf(entityType, BinaryData.class)) {
+            // Mono<BinaryData>
+            // The raw response is directly used to create an instance of BinaryData which then provides
+            // different methods to read the response. The reading of the response is delayed until BinaryData
+            // is read and depending on which format the content is converted into, the response is not necessarily
+            // fully copied into memory resulting in lesser overall memory usage.
+            if (methodParser.getReturnType().equals(StreamResponse.class)) {
+                // TODO (kasobol-msft) this is a hack.
+                // We don't need entity in that case but we can't change the else case yet
+                // it somehow relies on eager consumption and tests hang otherwise.
+                asyncResult = Mono.empty();
+            } else {
+                asyncResult = BinaryData.fromFlux(response.getSourceResponse().getBody());
+            }
```

Comment:
Let's revert this back to 
https://github.com/Azure/azure-sdk-for-java/blob/6628bb62b653fe2ec14889f181ba84675c57cee8/sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java#L498-L503

These changes will go separately when we work on StreamResponse.